velox/docs/develop/operators.rst

```diff
@@ -291,6 +292,125 @@ Note: Presto allows multiple global grouping sets in a single SQL query.
+
+To illustrate how ExpandNode works lets examine the following SQL query:
+
+.. code-block:: sql
+
+  SELECT l_orderkey, l_partkey, sum(l_suppkey) FROM lineitem GROUP BY ROLLUP(l_orderkey, l_partkey);
+
+In the planning phase, Spark generates an Expand operator with the following projection list:
+
+.. code-block::
+
+  [l_suppkey, l_orderkey, l_partkey, 0],
+  [l_suppkey, l_orderkey, null,      1],
+  [l_suppkey, null,       null,      3]
+
+Note: The grouping_id_0 in Spark is calculated based on a bitmask. If a certain column is selected, the bit value is assigned as 0; otherwise, it is assigned as 1. Therefore, the binary representation of the first row is (000), resulting in 0. The binary representation of the second row is (001), resulting in 1. The binary representation of the third row is (011), resulting in 3.
+
+For example, if the input rows are:
+
+.. code-block::
+
+  l_suppkey l_orderkey l_partkey
+  93        1          673
+  75        2          674
+  38        3          22
+
+After the computation by the ExpandNode, each row will generate 3 rows of data. So there will be a total of 9 rows:
+
+.. code-block::
+
+  l_suppkey l_orderkey l_partkey grouping_id_0 
+  93        1          673       0
+  93        1          null      1
+  93        null       null      3
+  75        2          674       0
+  75        2          null      1
+  75        null       null      3
+  38        3          22        0
+  38        3          null      1
+  38        null       null      3
+
+Then, these 9 rows of data are grouped by (l_orderkey, l_partkey, grouping_id_0), and the result of calculating l_orderkey, l_partkey, count(l_suppkey) is:
+
+.. code-block::
+
+  l_orderkey l_partkey count(l_suppkey)
+  1          673       1
+  null       null      3
+  1          null      1
+  2          null      1
+  2          674       1
+  3          null      1
+  3          22        1
+
+The following SQL query explain how the COUNT DISTINCT works with Expand in Spark:
+
+.. code-block:: sql
+
+  SELECT COUNT(DISTINCT l_suppkey), COUNT(DISTINCT l_partkey) FROM lineitem;
+
+In the planning phase, Spark generates an Expand operator with the following projection list:
+
+.. code-block::
+
+  [l_suppkey, null,      1],
+  [null,      l_partkey, 2]
+
+For example, if the input rows are:
+
+.. code-block::
+
+  l_suppkey l_partkey
+  93        673     
+  75        674      
+  38        22
+
+After the computation by the ExpandNode, each row will generate 2 rows of data. So there will be a total of 6 rows:
+
+.. code-block::
+
+  l_suppkey l_partkey grouping_id_0 
+  93        null      1
+  null      673       2
+  75        null      1
+  null      674       2
+  38        null      1
+  null      22        2
+
+Then, these 6 rows of data are grouped by grouping_id_0, and the result of calculating COUNT(DISTINCT l_suppkey), COUNT(DISTINCT l_partkey) is:
```

Comment:
Aggregation operator that follows, groups these rows by (l_suppkey,  l_partkey, grouping_id_0) and produces:

...

Another Aggregation operator that follows, computes global count(l_suppkey) and count(l_partkey) producing final result:

...

Reply:
@mbasmanova Updated.

Reply:
@JkSelf No need to reply to every single comment. This generated way too many non-actionable e-mail notifications.